sdk/eventhub/azure-eventhub/azure/eventhub/_buffered_producer/_buffered_producer.py

```diff
@@ -0,0 +1,221 @@
+            **kwargs: Any
+    ):
+        self._buffered_queue = queue.Queue()
+        self._cur_buffered_len = 0
+        self._executor: ThreadPoolExecutor = executor
+        self._producer: EventHubProducer = producer
+        self._lock = RLock()
+        self._not_empty = Condition(self._lock)
+        self._not_full = Condition(self._lock)
+        self._max_buffer_len = max_buffer_length
+        self._max_concurrent_sends = max_concurrent_sends
+        self._max_concurrent_sends_semaphore = Semaphore(self._max_concurrent_sends)
+        self._max_wait_time = max_wait_time
+        self._on_success = self.enhanced_callback_decorator(on_success)
+        self._on_error = self.enhanced_callback_decorator(on_error)
+        self._last_send_time = None
+        self._running = False
+        self._cur_batch: Optional[EventDataBatch] = None
+        self._max_message_size_on_link = max_message_size_on_link
+        self._send_futures = WeakSet()
+        self._check_max_wait_time_future = None
+        self.partition_id = partition_id
+
+    def start(self):
+        with self._lock:
+            self._cur_batch = EventDataBatch(self._max_message_size_on_link)
+            self._buffered_queue.put(self._cur_batch)
+            self._running = True
+            if self._max_wait_time:
+                self._last_send_time = time.time()
+                self._check_max_wait_time_future = self._executor.submit(self.check_max_wait_time_worker)
+
+    def stop(self, flush=True, timeout=None):
+        self._running = False
+        if self._check_max_wait_time_future:
+            try:
+                self._check_max_wait_time_future.result(timeout)
+            except Exception as exc:
+                _LOGGER.warning(
+                    "Partition {} stopped with error {!r}".format(
+                        self.partition_id,
+                        exc
+                    )
+                )
+        if flush:
+            self.flush(timeout=timeout, raise_error=False)
+        else:
+            if self._cur_buffered_len:
+                _LOGGER.warning(
+                    "Shutting down Partition {}."
+                    " There are still {} events in the buffer which will got lost".format(
+                        self.partition_id,
+                        self._cur_buffered_len
+                    )
+                )
+        self._producer.close()
+
+    def put_events(self, events, timeout=None):
+        # Put single event or EventDataBatch into the queue.
+        # This method would raise OperationTimeout if the queue does not have enough space for the input and
+        # flush cannot finish in timeout.
+        with self._not_full:
+            try:
+                new_events_len = len(events)
+            except TypeError:
+                new_events_len = 1
+            timeout_time = time.time() + timeout if timeout else None
+
+            if self._max_buffer_len - self._cur_buffered_len < new_events_len:
```

Comment:
We don't seem to be using the `not_full` condition?

Reply:
no it's not used, originally I was thinking of awaiting `check_max_wait_time_worker` to flush and to notify not full, however, I find it's easier to perform flush within `put_events` to get room as `check_max_wait_time_worker` might wait for a while before flushing.

Reply:
here's it's used as a lock, but I think in the future, we could introduce more sophisticated back pressure control strategy.

Reply:
I have logged an issue for this: https://github.com/Azure/azure-sdk-for-python/issues/23909